syft/pkg/cataloger/internal/dependency/resolver.go

```diff
@@ -0,0 +1,107 @@
+package dependency
+
+import (
+	"sort"
+
+	"github.com/scylladb/go-set/strset"
+
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/pkg/cataloger/generic"
+)
+
+// Specification holds strings that indicate abstract resources that a package provides for other packages and
+// requires for itself. These strings can represent anything from file paths, package names, or any other concept
+// that is useful for dependency resolution within that packing ecosystem.
+type Specification struct {
+	// Provides holds a list of abstract resources that this package provides for other packages.
+	Provides []string
+
+	// Requires holds a list of abstract resources that this package requires from other packages.
+	Requires []string
+}
+
+// Specifier is a function that takes a package and extracts a Specification, describing resources
+// the package provides and needs.
+type Specifier func(pkg.Package) Specification
+
+// RelationshipResolver uses a Specifier to resolve relationships between packages based on generic dependency claims
+// and provider claims from any given package.
+type RelationshipResolver struct {
+	specifier Specifier
+}
+
+func NewRelationshipResolver(s Specifier) RelationshipResolver {
+	return RelationshipResolver{
+		specifier: s,
+	}
+}
```

Comment:
do these types need to be exported, or do they need to exist at all? could the Resolve function below just be something like `resolveDependencies(Specifier, pkgs)`?

Reply:
agreed, I'll get rid of the type altogether